File: skills/lasersell/examples/auto-sell/auto_sell.py
"""Auto-sell stream example.

Behavior:
- Connects to stream websocket and tracks positions.
- On each `exit_signal_with_tx`, signs and submits transaction.
- Keeps running after individual submission failures.

Before running:
- Replace API key and wallet placeholders.
- Set `keypair_path` to a local Solana keypair JSON file.
"""

from __future__ import annotations

import asyncio
import json
from pathlib import Path

from solders.keypair import Keypair

from lasersell_sdk.exit_api import ExitApiClient, prove_ownership
from lasersell_sdk.stream.client import StreamClient, StreamConfigure, StrategyConfigBuilder
from lasersell_sdk.stream.session import StreamSession
from lasersell_sdk.tx import SendTargetHeliusSender, send_transaction, sign_unsigned_tx


async def main() -> None:
    api_key = "REPLACE_WITH_API_KEY"
    wallet_pubkeys = [
        "REPLACE_WITH_WALLET_PUBKEY_1",
        "REPLACE_WITH_WALLET_PUBKEY_2",
    ]
    keypair_path = Path("REPLACE_WITH_KEYPAIR_PATH")
    close_after_submit = False

    signer = read_keypair_file(keypair_path)

    # Register wallet (required for stream connection)
    api_client = ExitApiClient.with_api_key(api_key)
    proof = prove_ownership(signer)
    await api_client.register_wallet(proof)

    client = StreamClient(api_key)
    session = await StreamSession.connect(
        client,
        StreamConfigure(
            wallet_pubkeys=wallet_pubkeys,
            strategy=StrategyConfigBuilder()
                .target_profit_pct(5.0)
                .stop_loss_pct(1.5)
                .take_profit_levels([
                    {"profit_pct": 3, "sell_pct": 30, "trailing_stop_pct": 0},
                    {"profit_pct": 5, "sell_pct": 100, "trailing_stop_pct": 1},
                ])
                .build(),
            deadline_timeout_sec=45,
            send_mode="helius_sender",
            tip_lamports=1000,
        ),
    )

    pending_submissions: set[asyncio.Task[None]] = set()

    while True:
        event = await session.recv()
        if event is None:
            break

        if event.type == "position_opened" and event.handle is not None:
            print(
                "tracked position "
                f"wallet={event.handle.wallet_pubkey} "
                f"mint={event.handle.mint} "
                f"token_account={event.handle.token_account}"
            )
            continue

        if event.type == "exit_signal_with_tx" and event.message.get("type") == "exit_signal_with_tx":
            message = event.message
            task = asyncio.create_task(
                submit_exit_signal(
                    position_id=int(message["position_id"]),
                    wallet_pubkey=str(message["wallet_pubkey"]),
                    mint=str(message["mint"]),
                    unsigned_tx_b64=str(message["unsigned_tx_b64"]),
                    signer=signer,
                    close_after_submit=close_after_submit,
                    session=session,
                )
            )
            pending_submissions.add(task)
            task.add_done_callback(pending_submissions.discard)
            continue

        if event.type == "message" and event.message.get("type") == "error":
            print(
                "stream error "
                f"code={event.message.get('code')} "
                f"message={event.message.get('message')}"
            )

    await asyncio.gather(*pending_submissions, return_exceptions=True)
    raise RuntimeError("stream ended unexpectedly")


async def submit_exit_signal(
    position_id: int,
    wallet_pubkey: str,
    mint: str,
    unsigned_tx_b64: str,
    signer: Keypair,
    close_after_submit: bool,
    session: StreamSession,
) -> None:
    try:
        signed_tx = sign_unsigned_tx(unsigned_tx_b64, signer)
        signature = await send_transaction(SendTargetHeliusSender(), signed_tx)

        if close_after_submit:
            session.sender().close_by_id(position_id)

        print(f"submitted exit tx signature={signature} wallet={wallet_pubkey} mint={mint}")
    except Exception as error:
        print(
            "submission error "
            f"position_id={position_id} wallet={wallet_pubkey} mint={mint}: {error}"
        )


def read_keypair_file(path: Path) -> Keypair:
    raw = path.read_text(encoding="utf-8")
    parsed = json.loads(raw)
    if not isinstance(parsed, list):
        raise ValueError("keypair file must be an array of numbers")

    secret = bytes(int(value) for value in parsed)
    return Keypair.from_bytes(secret)


if __name__ == "__main__":
    asyncio.run(main())

File: skills/lasersell/examples/auto-sell/auto_sell.rs
//! Auto-sell stream example.
//!
//! This program connects to the LaserSell stream, listens for
//! `ExitSignalWithTx` events, signs each unsigned transaction with a local
//! keypair, and submits it through Helius Sender.
//!
//! Before running:
//! - Replace the API key and wallet placeholders below.
//! - Point `keypair_path` to a local Solana keypair file.
//!
//! Behavior notes:
//! - Transaction submission failures are logged and the stream keeps running.
//! - If the stream disconnects, the example returns an error.

use std::error::Error;

use lasersell_sdk::exit_api::{prove_ownership, ExitApiClient};
use lasersell_sdk::stream::client::{StrategyConfigBuilder, StreamClient, StreamConfigure};
use lasersell_sdk::stream::proto::{ServerMessage, TakeProfitLevelMsg};
use lasersell_sdk::stream::session::{StreamEvent, StreamSession};
use lasersell_sdk::tx::{send_transaction, sign_unsigned_tx, SendTarget};
use secrecy::SecretString;
use solana_sdk::signature::read_keypair_file;
use tokio::task::JoinSet;

struct SubmittedExit {
    wallet_pubkey: String,
    mint: String,
    signature: String,
}

fn handle_submission_result(result: Result<Result<SubmittedExit, String>, tokio::task::JoinError>) {
    match result {
        Ok(Ok(submission)) => {
            println!(
                "submitted exit tx signature={} wallet={} mint={}",
                submission.signature, submission.wallet_pubkey, submission.mint
            );
        }
        Ok(Err(message)) => {
            eprintln!("submission error: {message}");
        }
        Err(error) => {
            eprintln!("submission task failed: {error}");
        }
    }
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn Error>> {
    let api_key = "REPLACE_WITH_API_KEY".to_string();
    let wallet_pubkeys = vec![
        "REPLACE_WITH_WALLET_PUBKEY_1".to_string(),
        "REPLACE_WITH_WALLET_PUBKEY_2".to_string(),
    ];
    let keypair_path = "REPLACE_WITH_KEYPAIR_PATH".to_string();
    let close_after_submit = false;

    let keypair = read_keypair_file(keypair_path)?;
    let http = reqwest::Client::builder().build()?;

    // Register wallet (required for stream connection)
    let api_client = ExitApiClient::with_api_key(SecretString::new(api_key.clone()))?;
    let proof = prove_ownership(&keypair);
    api_client.register_wallet(&proof, None).await?;

    let client = StreamClient::new(SecretString::new(api_key));
    let configure = StreamConfigure {
        wallet_pubkeys,
        strategy: StrategyConfigBuilder::new()
            .target_profit_pct(5.0)
            .stop_loss_pct(1.5)
            .take_profit_levels(vec![
                TakeProfitLevelMsg { profit_pct: 3.0, sell_pct: 30.0, trailing_stop_pct: 0.0 },
                TakeProfitLevelMsg { profit_pct: 5.0, sell_pct: 100.0, trailing_stop_pct: 1.0 },
            ])
            .build(),
        deadline_timeout_sec: 45,
        send_mode: Some("helius_sender".to_string()),
        tip_lamports: Some(1000),
    };

    let mut session = StreamSession::connect(&client, configure).await?;
    let mut submissions = JoinSet::new();

    while let Some(event) = session.recv().await {
        match event {
            StreamEvent::PositionOpened { handle, .. } => {
                println!(
                    "tracked position wallet={} mint={} token_account={}",
                    handle.wallet_pubkey, handle.mint, handle.token_account
                );
            }
            StreamEvent::ExitSignalWithTx {
                handle: _,
                message:
                    ServerMessage::ExitSignalWithTx {
                        position_id,
                        wallet_pubkey,
                        mint,
                        unsigned_tx_b64,
                        ..
                    },
            } => {
                let signed_tx = sign_unsigned_tx(&unsigned_tx_b64, &keypair)?;
                let http = http.clone();
                let sender = session.sender();
                submissions.spawn(async move {
                    let signature = send_transaction(&http, &SendTarget::HeliusSender, &signed_tx).await.map_err(
                        |error| {
                            format!(
                                "send failed position_id={position_id} wallet={wallet_pubkey} mint={mint}: {error}"
                            )
                        },
                    )?;

                    if close_after_submit {
                        sender.close_by_id(position_id).map_err(|error| {
                            format!("close failed position_id={position_id}: {error}")
                        })?;
                    }

                    Ok::<SubmittedExit, String>(SubmittedExit {
                        wallet_pubkey,
                        mint,
                        signature,
                    })
                });
            }
            StreamEvent::Message(ServerMessage::Error { code, message }) => {
                eprintln!("stream error code={code} message={message}");
            }
            _ => {}
        }

        while let Some(result) = submissions.try_join_next() {
            handle_submission_result(result);
        }
    }

    while let Some(result) = submissions.join_next().await {
        handle_submission_result(result);
    }

    Err("stream ended unexpectedly".into())
}
